framework/src/kernels/ADKernelSUPG.C

```diff
@@ -0,0 +1,182 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADKernelSUPG.h"
+#include "MathUtils.h"
+#include "Assembly.h"
+
+// libmesh includes
+#include "libmesh/threads.h"
+
+#define PGParams                                                                                   \
+  params.addParam<MaterialPropertyName>(                                                           \
+      "tau_name", "tau", "The name of the stabilization parameter tau");                           \
+  params.addRequiredCoupledVar("velocity", "The velocity variable")
+
+defineADValidParams(ADKernelSUPG, ADKernel, PGParams;);
+defineADValidParams(ADVectorKernelSUPG, ADVectorKernel, PGParams;);
+
+template <typename T, ComputeStage compute_stage>
+ADKernelSUPGTempl<T, compute_stage>::ADKernelSUPGTempl(const InputParameters & parameters)
+  : ADKernelTempl<T, compute_stage>(parameters),
+    _tau(adGetADMaterialProperty<Real>("tau_name")),
+    _velocity(adCoupledVectorValue("velocity"))
+{
+}
+
+template <typename T, ComputeStage compute_stage>
+void
+ADKernelSUPGTempl<T, compute_stage>::computeResidual()
+{
+  prepareVectorTag(_assembly, _var.number());
+
+  precalculateResidual();
+  const unsigned int n_test = _grad_test.size();
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    const auto value = precomputeQpStrongResidual() * _ad_JxW[_qp] * _ad_coord[_qp] * _tau[_qp];
```

Comment:
shouldn't the tau be on the other term?

Reply:
You mean inside the `_i` loop? It's only a function of `_qp`

Reply:
Oh, I see. I thought SUPG is

(test + tau * grad_test * v * adv) * strong_residual

But I'm not an SUPG expert by any stretch of the definition of the word "expert" :-D

Reply:
> (test + tau * grad_test * v * adv) * strong_residual

This is almost correct. For a general transient advection-diffusion problem, 

u_dot + v * grad_u + grad * (-kappa * grad_u) - f = 0

the overall weak form would look like:

(test, u_dot) + (test, v * grad_u) + (grad_test, kappa * grad_u) - (test, f) - <test, n * kappa * grad_u> + (tau * v * grad_test, u_dot + v * grad_u + grad * (-kappa * grad_u) - f) = 0

So the difference is in the treatment of second order terms...those terms are integrated by parts when applying the standard test function, but not when applying the Petrov-Galerkin modificiation, tau * v * grad_test.

So in this design, an `ADKernelSUPG` derived object is only meant to be used to add the contribution from the Petrov-Galerkin modification. Other kernels should supply the standard test function contributions.